### PR TITLE
Fix profile penumbra part deux

### PR DESCRIPTION
**Pull request by randlet**
This is a replacement for PR #324 and includes a fix for LeedsTOR which broke due to improper peak detection.

Warning: I don't 100% understand the difference between threshold & x in `find_fwxm_peaks` and I need to investigate more whether it breaks anything else.  Any thoughts?

### CODE REVIEW

**Comment by randlet** [comment]
Not sure why yet but this PR breaks the starshot demo.

**Comment by jrkerns** [comment]
Let me know if these help: https://github.com/jrkerns/pylinac/blob/265ca6a41cb55fc0a696c3a4a9600822a4b230ea/pylinac/core/profile.py#L616-L619
and 
https://github.com/jrkerns/pylinac/blob/265ca6a41cb55fc0a696c3a4a9600822a4b230ea/pylinac/core/profile.py#L558-L564

**Comment by randlet** [comment]
I read those but somehow confused myself. Yes that makes sense. Thanks.

**Comment by randlet** [comment]
With this PR the two outside peaks don't get detected correctly.

starshot-peaks

**Comment by jrkerns** [comment]
> Not sure why yet but this PR breaks the starshot demo.

Looking at the starshot both with your recent change and this change now. I'm also doing some small refactoring.

**Comment by jrkerns** [comment]
FYI the multiple image combine test has been failing for a while.

**Comment by jrkerns** [comment]
Problem is somewhere in here: https://github.com/jrkerns/pylinac/blob/265ca6a41cb55fc0a696c3a4a9600822a4b230ea/pylinac/core/profile.py#L641-L646

**Comment by randlet** [comment]
Just came to exact same conclusion!  The first subprofile has this shape:

subprof

and the new penumbra point detection method is picking up the right hand side.  Should the penumbra detection take the gradient of the profile into consideration?  So when looking for a profile from RIGHT it should only consider points where gradient is negative?

**Comment by jrkerns** [comment]
So the new code is finding the FWXM based on the indices found, and when you have a profile like this then it will use the furthest-out index. So the FWXM has a right side of 4800. 
Figure_1

**Comment by randlet** [comment]
lol

**Comment by jrkerns** [comment]
> Just came to exact same conclusion! The first subprofile has this shape:
> 
> subprof
> 
> and the new penumbra point detection method is picking up the right hand side. Should the penumbra detection take the gradient of the profile into consideration? So when looking for a profile from RIGHT it should only consider points where gradient is negative?

I've not thought of using the sign of the gradient. That's pretty smart. I think this is why I did the center-looking-outward approach.

**Comment by jrkerns** [comment]
In the find_peaks function there's a parameter ``search_region``. This helps cut the left and right sides of the profile to avoid these cases. Unfortunately, I've not implemented this for FWXM methods.

**Comment by jrkerns** [comment]
https://github.com/jrkerns/pylinac/blob/265ca6a41cb55fc0a696c3a4a9600822a4b230ea/pylinac/core/profile.py#L572-L575

**Comment by jrkerns** [comment]
Hacky, but the smallest solution instead of passing through ``search_region`` everywhere is to cut off the left and right ends within ``subdivide`` to say, 80% width: https://github.com/jrkerns/pylinac/blob/265ca6a41cb55fc0a696c3a4a9600822a4b230ea/pylinac/core/profile.py#L670-L672

**Comment by randlet** [comment]
Check commit I just pushed

File: pylinac/planar_imaging.py [hunk 1]
```diff
@@ -738,7 +738,7 @@ def _phantom_angle_calc(self) -> float:
 
         start_angle_deg = self._determine_start_angle_for_circle_profile()
         circle = self._circle_profile_for_phantom_angle(start_angle_deg)
-        peak_idx = circle.find_fwxm_peaks(threshold=0.6, max_number=1)[0]
+        peak_idx = circle.find_fwxm_peaks(threshold=0.6, x=80, max_number=1)[0]
```
Comment (jrkerns) [inline pylinac/planar_imaging.py:741]: This x=80 is failing a number of tests. Why are you choosing 80?

Comment (randlet) [reply pylinac/planar_imaging.py:741]: No good reason...I think that I was somewhat confused and was experimenting with various parameters. (I'm not proud :frowning_face:  )

**Comment by randlet** [comment]
Yes that's ok with me.

**Comment by jrkerns** [comment]
Awesome. This is perfect. It will find peaks and also the penumbra points. Refactor....

**Comment by randlet** [comment]
Always nice when you can eliminate a heap of code :)

**Comment by randlet** [comment]
OK, I went back and actually figured out why the code wasn't working instead of sticking in an arbitrary x=80 :roll_eyes: . Thank you for calling me out on that!

The issue was the penumbra point detection still picked the wrong location if there were multiple places in the profile with the correct gradient and the correct threshold. I fixed it so that the location closest to the peak meeting those conditions is used and it all seems good now!  This is maybe just academic at this point if you're factoring most of this code out now, but I needed a fix in the interim.

**Comment by jrkerns** [comment]
I am refactoring it 🤷‍♂️. The scipy implementation is very good.

**Comment by randlet** [comment]
See if you can talk them into implementing phantom angle detection next :laughing: !

**Comment by jrkerns** [comment]
time to launch a Kaggle contest for it! 😂

**Comment by SimonBiggs** [comment]
> See if you can talk them into implementing phantom angle detection next 😆 !

If there's interest in applying a similar logic for phantom angle detection, here is how I detect field image rotation for WLutz work:

https://github.com/pymedphys/pymedphys/blob/3738e899c9944bb592079ce017c04e07abd78226/lib/pymedphys/_wlutz/findfield.py#L197-L220

**Comment by SimonBiggs** [comment]
> See if you can talk them into implementing phantom angle detection next laughing !

Probably a better and more clear example of using scipy's basinhopping to determine rotation:

https://github.com/pymedphys/pymedphys/blob/3738e899c9944bb592079ce017c04e07abd78226/lib/pymedphys/_electronfactors/core.py#L384-L422

**Comment by randlet** [comment]
Thanks! I'm implementing a PTW EPID QC phantom now so maybe I'll give that a shot.

**Pull request closed without merging**